File: lib/strategy/greedy.dart
import 'dart:math';

class EpsilonGreedy {
  final int n;
  final List<(int, double)> hatQs;
  final double epsilon;
  final Random random;

  EpsilonGreedy(this.n, {this.epsilon = 0.1, int? seed})
    : hatQs = List.filled(n, (0, 1.0)),
      random = Random(seed);

  int call(int reward, int action) {
    hatQs[action] = () {
      final counts = hatQs[action].$1 + 1;
      return (
        counts,
        // 期望增量公式，与直接求期望一样
        hatQs[action].$2 + 1.0 / counts * (reward - hatQs[action].$2),
      );
    }();

    return random.nextDouble() < epsilon
        ? random.nextInt(n)
        : argMax(hatQs.map((e) => e.$2));
  }
}

int argMax(Iterable<double> xs) {
  var result = (-1, 0.0);
  for (final times in xs.indexed) {
    if (times.$2 > result.$2) {
      result = times;
    }
  }
  return result.$1;
}

File: python/hands_on_rl/greedy.py
from typing import List, Tuple, Optional, Iterable
import numpy as np


class EpsilonGreedy:
    """Epsilon-greedy strategy for multi-armed bandits."""

    def __init__(self, n: int, epsilon: float = 0.1):
        self.n = n
        self.epsilon = epsilon
        self.hat_qs: List[Tuple[int, float]] = [(0, 1.0) for _ in range(n)]

    def __call__(self, reward: int, action: int) -> int:
        self.hat_qs[action] = self.calc_hat_q(reward, action)

        if np.random.random() < self.epsilon:
            return np.random.randint(0, self.n - 1)
        else:
            return int(np.argmax(np.array([q for _, q in self.hat_qs])))

    def calc_hat_q(self, reward: int, action: int) -> Tuple[int, float]:
        count, old_estimate = self.hat_qs[action]
        new_count = count + 1
        new_estimate = old_estimate + (1.0 / new_count) * (reward - old_estimate)
        return new_count, new_estimate


class DecayingEpsilonGreedy:
    """Epsilon-greedy strategy for multi-armed bandits."""

    def __init__(self, n: int):
        self.n = n
        self.hat_qs: List[Tuple[int, float]] = [(0, 1.0) for _ in range(n)]

    def __call__(self, reward: int, action: int) -> int:
        count, old_estimate = self.hat_qs[action]
        new_count = count + 1
        new_estimate = old_estimate + (1.0 / new_count) * (reward - old_estimate)
        self.hat_qs[action] = (new_count, new_estimate)

        if np.random.random() < 1.0 / new_count:
            return np.random.randint(0, self.n - 1)
        else:
            return int(np.argmax(np.array([q for _, q in self.hat_qs])))
